common/optimiser/Optimiser.java

```diff
@@ -45,30 +53,49 @@ public class Optimiser {
     private MPSolver solver;
     private MPSolverParameters parameters;
     private Status status;
-    private boolean hasSolver;
-    private double objectiveValue;
+    private boolean isConstraintsUpToDate;
+    private Double objectiveValue;
 
     public Optimiser() {
         variables = new ArrayList<>();
         constraints = new HashSet<>();
         objectiveCoefficients = new HashMap<>();
         status = Status.NOT_SOLVED;
-        hasSolver = false;
+        objectiveValue = null;
+        isConstraintsUpToDate = false;
     }
 
     public synchronized Status optimise(long timeLimitMillis) {
-        assert hasSolver;
         if (isOptimal()) return status;
+        assert solutionSatisfiesModel();
+        maySetSolver();
+        setSolutionAsHints();
         solver.setTimeLimit(timeLimitMillis);
         status = Status.of(solver.solve(parameters));
-        recordValues();
+        recordSolverValues();
         if (isOptimal()) releaseSolver();
+        assert solutionSatisfiesModel();
         return status;
     }
 
-    private void recordValues() {
-        objectiveValue = solver.objective().value();
-        if (status != Status.NOT_SOLVED && status != Status.ERROR) variables.forEach(OptimiserVariable::recordSolutionValue);
+    private void maySetSolver() {
+        if (solver == null) createSolver();
+        else if (isConstraintsUpToDate) {
+            solver.reset();
+            setConstraintCoefficients();
+        }
+    }
+
+    private void recordSolverValues() {
+        if ((isFeasible() || isOptimal()) && solver.objective().value() < objectiveValue()) {
+            objectiveValue = solver.objective().value();
+            variables.forEach(OptimiserVariable::recordSolutionValue);
+        }
```

Comment:
we now only record the objective and solutions to optimiser variables whenever a solution is found that is actually better than the prior one!
We require this because we aren't using a "true" incremental solver, but restarting each time with the hints from the prior one.

Reply:
If the solver is deterministic, we might just be stuck in a Sisyphean loop here. Which isn't the problem with SAT solver, to my knowledge.

Reply:
it's true - i'm thinking once we can background plan we can incrementally give it more time if it doesn't make progress somehow?